### PR TITLE
[CELEBORN-1036][FOLLOWUP] totalInflightReqs should decrement when batchIdSet contains the batchId to avoid duplicate caller of removeBatch

### PR DESCRIPTION
**Pull request by SteNicholas**
### What changes were proposed in this pull request?

`totalInflightReqs` decrements when `batchIdSet` contains the `batchId` to avoid duplicate caller of `removeBatch` in `InFlightRequestTracker`.

### Why are the changes needed?

Caller of `InFlightRequestTracker#removeBatch` may be duplicated, which cause that `totalInflightReqs` could be negative. The source of truth should be that `totalInflightReqs` should decrement when `batchIdSet` contains the `batchId`. If `batchIdSet` does not contain the `batchId`, it does not need to decrement `totalInflightReqs`.

```
23/12/05 20:05:01 [Executor task launch worker for task 17.0 in stage 10.0 (TID 206)] ERROR InFlightRequestTracker: After waiting for 1200000 ms, there are still -1 batches in flight for hostAndPushPort [], which exceeds the current limit 0.
23/12/05 20:05:01 [Executor task launch worker for task 17.0 in stage 10.0 (TID 206)] WARN InFlightRequestTracker: Clear InFlightRequestTracker
23/12/05 20:05:01 [Executor task launch worker for task 17.0 in stage 10.0 (TID 206)] ERROR Executor: Exception in task 17.0 in stage 10.0 (TID 206)
org.apache.celeborn.common.exception.CelebornIOException: Waiting timeout for task 4-17-0 while limiting zero in-flight requests
	at org.apache.celeborn.client.ShuffleClientImpl.limitZeroInFlight(ShuffleClientImpl.java:598)
	at org.apache.celeborn.client.ShuffleClientImpl.prepareForMergeData(ShuffleClientImpl.java:1175)
	at org.apache.spark.shuffle.celeborn.HashBasedShuffleWriter.close(HashBasedShuffleWriter.java:455)
	at org.apache.spark.shuffle.celeborn.HashBasedShuffleWriter.write(HashBasedShuffleWriter.java:210)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:100)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:141)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:589)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1545)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:594)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Internal tests.

### CODE REVIEW

**Comment by SteNicholas** [comment]
@waitinfuture, @onebox-li, @AngersZhuuuu, could you help to review this follow up to improve the batch operation for consistency of totalInflightReqs, which avoid to duplicate adding or removing batch?

File: common/src/main/java/org/apache/celeborn/common/write/InFlightRequestTracker.java [hunk 1]
```diff
@@ -173,6 +169,10 @@ public boolean limitZeroInFlight() throws IOException {
     return times <= 0;
   }
 
+  private int totalInflightReqs() {
+    return inflightBatchesPerAddress.values().parallelStream().mapToInt(Set::size).sum();
```
[the remark this answers is not in the document]
Comment (SteNicholas) [reply common/src/main/java/org/apache/celeborn/common/write/InFlightRequestTracker.java:173]: @FMX, how expensive does getting the count?

File: common/src/main/java/org/apache/celeborn/common/write/InFlightRequestTracker.java [hunk 2]
```diff
@@ -58,11 +55,10 @@ public InFlightRequestTracker(CelebornConf conf, PushState pushState) {
...
-    Set<Integer> batchIdSetPerPair =
+    Set<Integer> batchIdSet =
         inflightBatchesPerAddress.computeIfAbsent(
             hostAndPushPort, id -> ConcurrentHashMap.newKeySet());
-    batchIdSetPerPair.add(batchId);
-    totalInflightReqs.increment();
+    batchIdSet.add(batchId);
   }
 
   public void removeBatch(int batchId, String hostAndPushPort) {
```
[the remark this answers is not in the document]
Comment (FMX) [reply common/src/main/java/org/apache/celeborn/common/write/InFlightRequestTracker.java:64]: I mean that you can invoke `totalInflightReqs.decrement()` if the `batchIdSet` is not null.

**Comment by SteNicholas** [comment]
@FMX, thanks for detailed review. I have addressed above comments. PTAL.

**Review by FMX — APPROVED** [review]
LGTM. Thanks. Merged into main and 0.4.

**Pull request closed without merging**